emission/tests/modellingTests/TestGreedySimilarityBinning.py

```diff
@@ -24,31 +30,414 @@ def testBinning(self):
         # within a radius that should have them binned.
         n = 20
         m = 5
+        
+        # trip_part: when mock trips are generated, coordinates of this part of 
+        #            m trips will be within the threshold. trip_part can take one
+        #            among the four values:
+        #
+        #            1. '__' ->(None, meaning NEITHER origin nor destination of any trip will lie 
+        #             within the mentioned threshold when trips are generated),
+        #
+        #            2. 'o_' ->(origin, meaning ONLY origin of m trips will lie within the mentioned 
+        #             threshold when trips are generated),
+        #
+        #            3. '_d' ->(destination),meaning ONLY destination of m trips will lie within the 
+        #             mentioned threshold when trips are generated)
+        #
+        #            4. 'od' ->(origin and destination,meaning BOTH origin and destination of m trips
+        #             will lie within the mentioned threshold when trips are generated)
+
         trips = etmm.generate_mock_trips(
             user_id="joe", 
             trips=n, 
             origin=(0, 0), 
-            destination=(1, 1), 
+            destination=(1, 1),
+            trip_part='__',
             label_data=label_data, 
             within_threshold=m, 
             threshold=0.001,  # ~ 111 meters in degrees WGS84
         )
+    
 
-        # pass in a test configuration to the binning algorithm
-        model_config = {
+        # pass in a test configuration to the binning algorithm.
+        #
+        # clustering_way : Part of the trip used for checking pairwise proximity.
+        #                  Can take one of the three values:
+        #                  
+        #                   1. 'origin' -> using origin of the trip to check if 2 points
+        #                                   lie within the mentioned similarity_threshold_meters
+        #                   2. 'destination' -> using destination of the trip to check if 2 points
+        #                                       lie within the mentioned similarity_threshold_meters
+        #                   3. 'origin-destination' -> both origin and destination of the trip to check 
+        #                                             if 2 points lie within the mentioned 
+        #                                              similarity_threshold_meters
+        
+        model1_config = {
             "metric": "od_similarity",
-            "similarity_threshold_meters": 500,  # meters,
+            "similarity_threshold_meters": 111,  # meters,
```

Comment:
why did you change this to `111`?

Reply:
This was done to prevent pairing of points that were more than 111 meters away from each other, but less than 500 m. Now I realize that wont be possible, because the reference points of coordinate generation ( (0,0) or (1,1) ) are actually more than 100 kms away.

 However, after looking into this, keeping it to either (111 or 500) of the values MIGHT cause problem of wrong binning  because of the way these coordinates are generated in `modellingTestAssets.py` `generate_trip_coordinates` :
 

> ```radius = random.uniform(0, radius_threshold) if within_threshold else random.uniform(*radius_threshold, max) ```

This is why I feel this would create problem in certain cases :


 If we want  coordinates to be within `within_threshold` (set to 111m)  (points  in white like p1,p2,p3) , the radius is sampled to be less than `radius_threshold` , otherwise the radius is sampled in red marked area. However, say ,we generate 2 points (in blue) one within the `within_threshold` and the other one outside the `within_threshold` with the aim for no binning . However, our binning threshold d (500m) might result in binning them together if the outside point is sampled close enough to the circle.

 
![MicrosoftTeams-image (2)](https://github.com/e-mission/e-mission-server/assets/37581386/7d4a68ef-5a7b-4f26-afd1-d36e9914648d)

 
The way to fix this is when the points are not to be binned together, they need to be sampled around center points that are farther away than the binning threshold (500m). For eg, Below we take center points as (0,0) , (1,1) , (2,2). 
 
![MicrosoftTeams-image (1)](https://github.com/e-mission/e-mission-server/assets/37581386/c266682d-a37d-42b0-86fb-a82c7d8ce63e)

 Fixing this in next commit.

Reply:
As long as the thresholds are consistent, I don't see any issues with 500 (and the test has been running, and passing)  on production for a while with 500.

Essentially, we create "fake" trips. if the ends of the trips are not within the expected bin, we discard the trips. So I don't understand how we would get trips that are outside the bin radius.

Reply:
For the record, the related issue was:
https://github.com/e-mission/e-mission-docs/issues/709
and the related PR for the first set of changes was:
https://github.com/e-mission/e-mission-server/pull/852

Reply:
Neither had any discussion around the choice of 500.
Having said that, even if we are not discarding "invalid" trips right now, it seems like we can discard invalid trips going forward, since we don't really care about what the trips are, and they are mock trips anyway. We just want to have "n" trips that match some characteristic.

By "invalid", I mean, not matching the kinds of mock trips that we want.
So if we want 5 trips, 3 of which are binned in both o-d and 2 of which are binned only in o,
we can oversample (overcreate mock trips) and discard the ones that don't match these criteria until the criteria are met.